sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.implementation.{CosmosClientMetadataCachesSnapshot, Strings}
+import com.azure.cosmos.models.{CosmosChangeFeedRequestOptions, FeedRange}
+import com.azure.cosmos.spark.CosmosPredicates.{assertNotNull, assertNotNullOrEmpty, assertOnSparkDriver, requireNotNull}
+import com.fasterxml.jackson.databind.node.ObjectNode
+import org.apache.spark.broadcast.Broadcast
+import reactor.core.scala.publisher.SMono
+import reactor.core.scala.publisher.SMono.PimpJMono
+import reactor.core.scheduler.Schedulers
+
+import java.time.Instant
+import java.util.{Timer, TimerTask}
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+import scala.collection.concurrent.TrieMap
+
+// The partition metadata here is used purely for a best effort
+// estimation of number of Spark partitions needed for a certain
+// physical partition - it is not guaranteeing functional correctness
+// because the cached metadata could be old or even be for a different
+// container after deletion and recreation of a container
+private object PartitionMetadataCache extends CosmosLoggingTrait {
+  private[this] val Nothing = 0
+  private[this] val cache = new TrieMap[String, PartitionMetadata]
+  private[this] var cacheTestOverride: Option[TrieMap[String, PartitionMetadata]] = None
```

Comment:
I definitely prefer test hooks over use of reflection - and that is the rule of thumb we have been using in .Net V3 (as long as test hooks don't become public surface area). private internal access won't help because the core data structures could not be final/read-only in that case. And that was the motivation to add the additional test hooks. I wanted to reserve the immutability as long as not under test. But maybe there are more elegant options - haven't spent much time on it - will add a comment to go over it again.

Reply:
I personally prefer to not add test hooks in the class if possible. But I guess it is not always possible/easy. Using test hooks definitely is better than having lesser test coverage. thanks for explanation. LGTM.